tracing-subscriber/tests/option_filter_interest_caching.rs

```diff
@@ -0,0 +1,52 @@
+// A separate test crate for `Option<Filter>` for isolation from other tests
+// that may influence the interest cache.
+
+use std::sync::{Arc, atomic::{AtomicUsize, Ordering}};
+use tracing_mock::{subscriber, expect};
+use tracing_subscriber::{filter, prelude::*, Subscribe};
+
+/// A `None` filter should always be interested in events, and it should not
+/// needlessly degrade the caching of other filters.
+#[test]
+fn none_interest_cache() {
+    let (subscribe_none, handle_none) = subscriber::mock()
+        .event(expect::event())
+        .event(expect::event())
+        .only()
+        .run_with_handle();
+    let subscribe_none = subscribe_none
+        .with_filter(None::<filter::DynFilterFn<_>>);
+
+    let times_filtered = Arc::new(AtomicUsize::new(0));
+    let (subscribe_filter_fn, handle_filter_fn) = subscriber::mock()
+        .event(expect::event())
+        .event(expect::event())
+        .only()
+        .run_with_handle();
+    let subscribe_filter_fn = subscribe_filter_fn
+        .with_filter(filter::filter_fn({
+            let times_filtered = Arc::clone(&times_filtered);
+            move |_| {
+                times_filtered.fetch_add(1, Ordering::Relaxed);
+                true
+            }
+        }));
+
+    let subscriber = tracing_subscriber::registry()
+        .with(subscribe_none)
+        .with(subscribe_filter_fn);
+
+    let _guard = subscriber.set_default();
+    for _ in 0..2 {
+        tracing::debug!(target: "always_interesting", x="bar");
+    }
+
+    // The `None` filter is unchanging and performs no filtering, so it should
+    // be cacheable and always be interested in events. The filter fn is a
+    // non-dynamic filter fn, which means the result can be cached per callsite.
+    // The filter fn should only need to be called once, and the `Option` filter
+    // should not interfere in the caching of that result.
+    assert_eq!(times_filtered.load(Ordering::Relaxed), 1);
```

Comment:
With `sometimes`, the filter fn is evaluated three times:
1. For the initial check to see if the callsite is enabled (effectively returns `Interest::always`)
2. For the first debug event
3. For the second debug event

With `always`, only #1 occurs. Both filters are always interested, so there's no need to reevaluate the filter fn.